crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -345,6 +345,80 @@ def services
       end
       handle_asynchronously :services
 
+      def openstackbackup
+        ::Crowbar::UpgradeStatus.new.start_step(:backup_openstack)
+
+        crowbar_lib_dir = "/var/lib/crowbar"
+        dump_path = "#{crowbar_lib_dir}/upgrade/6-to-7-openstack_dump.sql"
+        if File.exist?(dump_path)
+          Rails.logger.debug("OpenStack backup already exists. Skipping...")
+          return
+        end
+
+        psql = postgres_params
+        query = "SELECT SUM(pg_database_size(pg_database.datname)) FROM pg_database;"
+        cmd = "PGPASSWORD=#{psql[:pass]} psql -t -h #{psql[:host]} -U #{psql[:user]} -c '#{query}'"
+
+        Rails.logger.debug("Checking size of OpenStack database")
+        db_size = run_cmd(cmd)
+        unless db_size[:exit_code].zero?
+          Rails.logger.error(
+            "Failed to check size of OpenStack database: #{db_size[:stdout_and_stderr]}"
+          )
+          raise ::Crowbar::Error::UpgradeDatabaseSizeError.new(
+            db_size[:stdout_and_stderr]
+          )
+        end
+
+        free_space = run_cmd(
+          "LANG=C df -x 'tmpfs' -x 'devtmpfs' -B1 -l --output='avail' #{crowbar_lib_dir} | tail -n1"
+        )
+        unless free_space[:exit_code].zero?
+          Rails.logger.error("Cannot determine free disk space: #{free_space[:stdout_and_stderr]}")
+          raise ::Crowbar::Error::UpgradeFreeDiskSpaceError.new(
+            free_space[:stdout_and_stderr]
+          )
+        end
+        if free_space[:stdout_and_stderr].strip.to_i < db_size[:stdout_and_stderr].strip.to_i
```

Comment:
I'd add some extra pessimism factor to it (like 10% more than database size or something like that)..